flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adapter/DefaultExecutionTopology.java

```diff
@@ -159,50 +162,79 @@ public static DefaultExecutionTopology fromExecutionGraph(
+                logicalPipelinedRegionsByJobVertexId.put(vertex.getId(), logicalPipelinedRegion);
             }
         }
 
+        return logicalPipelinedRegionsByJobVertexId;
+    }
+
+    public void notifyExecutionGraphUpdated(
+            final DefaultExecutionGraph executionGraph,
+            final List<ExecutionJobVertex> newJobVertices) {
+
+        checkNotNull(executionGraph, "execution graph can not be null");
+
+        final Set<JobVertexID> newVertexSet =
+                newJobVertices.stream()
+                        .map(ExecutionJobVertex::getJobVertexId)
+                        .collect(Collectors.toSet());
+
+        // any PIPELINED input should be from within this new set so that existing pipelined regions
+        // will not change
+        newJobVertices.stream()
+                .map(ExecutionJobVertex::getJobVertex)
+                .flatMap(v -> v.getInputs().stream())
+                .map(JobEdge::getSource)
+                .filter(r -> r.getResultType().isPipelined())
+                .map(IntermediateDataSet::getProducer)
+                .map(JobVertex::getID)
+                .forEach(id -> checkState(newVertexSet.contains(id)));
+
+        final Iterable<ExecutionVertex> newAddedExecutionVertices =
+                newJobVertices.stream()
+                        .flatMap(jobVertex -> Stream.of(jobVertex.getTaskVertices()))
+                        .collect(Collectors.toList());
+
+        updateExecutionGraphIndex(newAddedExecutionVertices);
+
+        updatePipelinedRegions(newAddedExecutionVertices);
+
+        ensureCoLocatedVerticesInSameRegion(pipelinedRegions, executionGraph);
+    }
+
+    public static DefaultExecutionTopology fromExecutionGraph(
+            DefaultExecutionGraph executionGraph) {
+        checkNotNull(executionGraph, "execution graph can not be null");
+
+        EdgeManager edgeManager = executionGraph.getEdgeManager();
+
+        Iterable<ExecutionVertex> executionVertices = executionGraph.getAllExecutionVertices();
+
+        DefaultExecutionTopology schedulingTopology =
+                new DefaultExecutionTopology(
+                        () ->
+                                IterableUtils.toStream(executionVertices)
+                                        .map(ExecutionVertex::getID)
+                                        .collect(Collectors.toList()),
+                        edgeManager,
+                        computeLogicalPipelinedRegionsByJobVertexId(executionGraph));
+
+        schedulingTopology.notifyExecutionGraphUpdated(
+                executionGraph,
+                IterableUtils.toStream(executionGraph.getVerticesTopologically())
+                        .filter(ExecutionJobVertex::isInitialized)
+                        .collect(Collectors.toList()));
+
+        return schedulingTopology;
+    }
+
+    private void updateExecutionGraphIndex(Iterable<ExecutionVertex> executionVertices) {
```

Comment:
executionVertices -> newExecutionVertices

Reply:
> updateExecutionGraphIndex

The name is outdated.
Maybe rename it to `generateNewExecutionVerticesAndResultPartitions`, or any other good ideas?

Reply:
`generateNewExecutionVerticesAndResultPartitions ` is good for me, and I will also rename the`updatePipelinedRegions ` to `generateNewPipelinedRegions`